feat(task): Реализация сервиса управления контрагентами.

ChiniakinD opened the pull request:
Реализация RESTful веб-сервиса для управления данными о контрагентах, странах, отраслях и формах организаций.

JavaStarterCloud committed:
feat(task): Реализация сервиса управления контрагентами.
Реализация RESTful веб-сервиса для управления данными о контрагентах, странах, отраслях и формах организаций.

JavaStarterCloud committed:
fix(task): Изменен нейминг директории

Замена директории org на com

JavaStarterCloud committed:
test(task): Добавление тестов для контроллеров.

Для ContractorController, CountryController, IndustryController, OrgFormController добавлены тесты.

gobaz095 commented on src/main/java/com/chiniakin/config/JdbcConfig.java:
```diff
@@ -0,0 +1,40 @@
+     * @return DataSource для PostgreSQL.
+     */
+    @Bean
+    public DataSource dataSource() {
```
А почему зачем тогда конфиги в application.yaml?

ChiniakinD replied:
[528a890](https://github.com/ChiniakinD/contractor/pull/1/commits/528a89085d158072d793a7a341a9c0d3585ff505)

gobaz095 commented on src/main/java/com/chiniakin/config/mapper/ContractorMapper.java:
```diff
@@ -0,0 +1,37 @@
+     * @return Модель ContractorModel
+     */
+    public ContractorModel toModel(Contractor contractor) {
+        ContractorModel contractorModel = modelMapper.map(contractor, ContractorModel.class);
```
Ок, но надо быть очень аккуратным с автоматическими мапперами. Иногда даже стоит написать отдельно тесты.

gobaz095 commented on src/main/java/com/chiniakin/controller/ContractorController.java:
```diff
@@ -0,0 +1,100 @@
+     * @param size размер страницы.
+     * @return контрагентов.
+     */
+    @GetMapping("/search")
```
Для исключения путаницы лучше назвать /all

gobaz095 commented on src/main/java/com/chiniakin/service/ContractorServiceImpl.java:
```diff
@@ -0,0 +1,171 @@
+            sqlQuery.append(" and inn = ?");
+            params.add(contractorFilter.getInn());
+        }
+        if (contractorFilter.getOgrn() != null) {
```
Зависит от бизнес логики, но может не помешать isBlank

ChiniakinD replied:
https://github.com/ChiniakinD/contractor/pull/1/commits/6b77a090a3424f5115004cd0e8cbc5a030511f54

gobaz095 commented on src/main/java/com/chiniakin/service/ContractorServiceImpl.java:
```diff
@@ -0,0 +1,171 @@
+        sqlQuery.append("left join industry i on c.industry = i.id ");
+        sqlQuery.append("left join org_form o on c.org_form = o.id ");
+
+        sqlQuery.append("where 1=1");
```
Можно и тупо собрать where и если оно пустое не добавлять. Но и так и так пойдет.

gobaz095 commented on src/main/java/com/chiniakin/service/ContractorServiceImpl.java:
```diff
@@ -0,0 +1,171 @@
+    @Override
+    public Page<ContractorModel> searchWithNativeFilters(ContractorFilter contractorFilter, Pageable pageable) {
+        StringBuilder sqlQuery = new StringBuilder("SELECT c.*, co.name as country_name, i.name as industry_name, o.name as org_form_name FROM contractor c ");
+        List<Object> params = new ArrayList<>();
```
Чтобы уменьшить риск дальнейших ошибок при изменении этого метода, лучше использовать именованные переменные Map<String, Object> params или вообще MapSqlParameterSource
Вместе с [NamedParameterJdbcTemplate](https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/jdbc/core/namedparam/NamedParameterJdbcTemplate.html)

ChiniakinD replied:
https://github.com/ChiniakinD/contractor/pull/1/commits/27cfb03da5649a6ad2747c3b1117a298cb3cbf51

gobaz095 commented on src/main/java/com/chiniakin/service/ContractorServiceImpl.java:
```diff
@@ -0,0 +1,171 @@
+        params.add(pageable.getPageSize());
+        params.add(pageable.getOffset());
+        try {
+            List<Contractor> contractors = jdbcTemplate.query(sqlQuery.toString(), params.toArray(), (rs, rowNum) -> {
```
вынести (rs, rowNum) -> в отдельный маппер бы

ChiniakinD replied:
https://github.com/ChiniakinD/contractor/pull/1/commits/8ce7289cde05c2af0d9ee12e33d3ef4d444c3f6e

gobaz095 reviewed (commented):
И тесты хочу не только на слой контроллеров, но также хотя бы на слой репозиториев. Сервисы тоже нужны обычно. В данном случае, у тебя это все пересекается и отличается мало. Но чаще всего логика репозитори и сервиса отличается сильно, и контроллер тоже может иметь свою логику. Плюс мы тестируем немного разные вещи, DTO != entity. А на бы надо проверить что у нас именно интеграционные тесты с БД работают как надо. И сохраняется и возвращается одно и тоже из БД.

И на экспепшены тоже хочу тесты, раз они написаны.

gobaz095 reviewed (changes requested):
И не вижу интеграционных тестов с testcontainers.
И в задании был все же liquibase, а не flyway. Это более мощный и гибкий инструмент на данный момент.

JavaStarterCloud committed:
feat(): Удалил JdbcConfig.java

JavaStarterCloud committed:
feat(): Заменил flyway на liquibase

1. Заменил flyway на liquibase.
2. Изменил расположение V1__init_schema.sql и V2__insert.sql

JavaStarterCloud committed:
feat(ContractorServiceImpl): Добавлено использование isBlank

JavaStarterCloud committed:
feat(ContractorServiceImpl): Добавлен отдельный маппер ContractorRowMapper.

JavaStarterCloud committed:
feat(ContractorServiceImpl): Внедрено использование MapSqlParameterSource, NamedParameterJdbcTemplate

gobaz095 commented on src/main/java/com/chiniakin/service/ContractorServiceImpl.java:
```diff
@@ -57,50 +57,50 @@ public Page<ContractorModel> searchWithNativeFilters(ContractorFilter contractor
         if (!isBlank(contractorFilter.getName())) {
-            sqlQuery.append(" and name = ?");
-            params.add(contractorFilter.getName());
+            sqlQuery.append(" and name = name");
```
Ошибка забыл : 
тестами покрой)

gobaz095 commented on src/main/java/com/chiniakin/service/ContractorServiceImpl.java:
```diff
@@ -57,50 +57,50 @@ public Page<ContractorModel> searchWithNativeFilters(ContractorFilter contractor
         if (!isBlank(contractorFilter.getNameFull())) {
-            sqlQuery.append(" and name_full = ?");
-            params.add(contractorFilter.getNameFull());
+            sqlQuery.append(" and name_full = nameFull");
```
Ошибка забыл : 
тестами покрой)

gobaz095 commented on src/main/resources/liquibase/migration/V2__insert.sql:
Добавь пожалуйста к миграциям описание. например как тут, с автором и тд:
https://habr.com/ru/articles/251617/

ChiniakinD replied:
https://github.com/ChiniakinD/contractor/pull/1/commits/019430e589ab02c78c01854b063ef16f13e5f429

JavaStarterCloud committed:
fix(ContractorServiceImpl): Добавил символ ':' для параметров в запросах

JavaStarterCloud committed:
refactor(ContractorController): Изменил название пути запроса

JavaStarterCloud committed:
test(): Добавление тестов

1. Добавил тестирование CountryRepository, IndustryRepository, OrgFormRepository.
2. Добавил testcontainers.

JavaStarterCloud committed:
refactor(migration): Добавил описание к миграциям

gobaz095 commented on src/main/java/com/chiniakin/config/mapper/ContractorRowMapper.java:
```diff
@@ -0,0 +1,57 @@
+     */
+    public Contractor mapRow(ResultSet rs, int rowNum) throws SQLException {
+
+        Contractor contractor = new Contractor()
```
Верю что ты усвоил паттерн билдер:) Можешь использовать ломбоковский. @Builder

ChiniakinD replied:
https://github.com/ChiniakinD/contractor/pull/1/commits/b6341ff18612cce9edabdb63ed74a780cf9dadb1

gobaz095 commented on src/main/java/com/chiniakin/service/ContractorServiceImpl.java:
```diff
@@ -0,0 +1,139 @@
+    public void deleteContractorById(String id) {
+        Contractor contractor = contractorRepository.findByIdWithDetailsOrThrow(id);
+        contractor.setActive(false);
+        contractorRepository.save(contractor);
```
Не критично, но если мы говорим о производительности, то зачем нам тянуть целиком объект чтобы у него изменить одно поле? Легче сразу сделать update contractor set is_active = false where id = id

ChiniakinD replied:
https://github.com/ChiniakinD/contractor/pull/1/commits/d2708c2cac73e30ce5c2d1e8d67def19e0cc255c

gobaz095 commented on src/main/resources/liquibase/migration/V1__init_schema.sql:
В будущем попробуй yaml, он предпочтительнее из-за систематизации изменений, единообразия, большей выходной информации и тд.

gobaz095 commented on src/main/java/com/chiniakin/model/ContractorModel.java:
Про описание сваггера, на всякий случай, я ожидаю увидеть тут @Schema, и все прочее в контроллерах.

gobaz095 reviewed (approved):
Можно влить, описание swagger и прочее отдельным уже ПРом.

gobaz095 reviewed (approved):
ок можно слить

JavaStarterCloud committed:
refactor(ContractorRowMapper): Добавил использование аннотации @builder

JavaStarterCloud committed:
feat():
1. Добавил возможность получения только активных объектов.
2. Добавил тесты.

JavaStarterCloud committed:
refactor(delete методы):Оптимизировал логику удаления
